llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Transforms/Transforms.cpp

```diff
@@ -902,6 +904,109 @@ struct GeneralizePackOpPattern : OpRewritePattern<PackOp> {
+
+    Location loc = unpackOp.getLoc();
+    Attribute zeroIdxAttr = rewriter.getIndexAttr(0);
+    Attribute oneIdxAttr = rewriter.getIndexAttr(1);
+    SmallVector<OpFoldResult> readOffsets(inputRank, zeroIdxAttr);
+    SmallVector<OpFoldResult> readStrides(inputRank, oneIdxAttr);
+
+    auto mixedTiles = unpackOp.getMixedTiles();
+    SmallVector<OpFoldResult> readSizes(outputRank, oneIdxAttr);
+    readSizes.append(mixedTiles.begin(), mixedTiles.end());
+
+    // Explicitly create the type because the inner tile size could be 1. We
+    // want to represent the whole inner tile in this case.
+    SmallVector<int64_t> readShape;
+    for (auto sz : mixedTiles) {
+      auto cst = getConstantIntValue(sz);
+      readShape.push_back(cst ? cst.value() : ShapedType::kDynamicSize);
+    }
+    Type elemType = unpackOp.getInputType().getElementType();
+    auto readType = RankedTensorType::get(readShape, elemType);
+    Value innerTile = rewriter.create<tensor::ExtractSliceOp>(
+        loc, readType, unpackOp.getInput(), readOffsets, readSizes,
+        readStrides);
+
+    SmallVector<int64_t> innerDimsPos =
+        extractFromI64ArrayAttr(unpackOp.getInnerDimsPos());
+    auto interchangeVector =
+        computeInterchangeFromDimPos(innerDimsPos, outputRank);
+    SmallVector<int64_t> transpShape =
+        interchange<int64_t>(readShape, interchangeVector);
+    auto transpType = RankedTensorType::get(transpShape, elemType);
+
+    Value empty = rewriter.create<tensor::EmptyOp>(loc, transpShape, elemType);
+
+    // TODO(hanchung): use linalg.transpose op after bumping MLIR. There is a
+    // bug in linalg.tranpose op that is not landed to IREE yet. See
+    // https://github.com/llvm/llvm-project/issues/58713
+    int numLoops = transpShape.size();
+    SmallVector<StringRef> loopTypes(numLoops, getParallelIteratorTypeName());
+
+    SmallVector<AffineExpr> outputExprs = llvm::to_vector<4>(
+        llvm::map_range(interchangeVector, [&](int64_t index) -> AffineExpr {
+          return rewriter.getAffineDimExpr(index);
+        }));
+    SmallVector<AffineMap, 2> indexingMaps = {
+        AffineMap::getMultiDimIdentityMap(numLoops, rewriter.getContext()),
+        AffineMap::get(numLoops, 0, outputExprs, rewriter.getContext())};
+
+    auto transposedOp = rewriter.create<linalg::GenericOp>(
+        loc, empty.getType(), innerTile, empty, indexingMaps, loopTypes,
+        [&](OpBuilder &nestedBuilder, Location nestedLoc, ValueRange args) {
+          nestedBuilder.create<linalg::YieldOp>(nestedLoc, args[0]);
+        });
+    // Handle in-complete tiles.
+    SmallVector<OpFoldResult> tileStrides(numLoops, oneIdxAttr);
+    SmallVector<OpFoldResult> tileOffsets(numLoops, zeroIdxAttr);
+    SmallVector<OpFoldResult> tileSizes;
+    for (int dim : innerDimsPos) {
+      tileSizes.push_back(getDim(rewriter, loc, unpackOp.getOutput(), dim));
+    }
+    tileSizes = interchange<OpFoldResult>(tileSizes, interchangeVector);
```

Comment:
Similarly for `tileSizes` you should be able to use `PackOp::getResultShape`.

Reply:
The getResultShape is getting the packed result shape. That is a padded shape. What we need here is to extract exact tensor from the transposed padded shape.

E.g.,

```
 %0 iree_linalg_ext.unpack %input
  inner_dims_pos = [0, 1]
  inner_tiles = [8, 2]
  into %output : (tensor<1x1x8x2xf32> tensor<5x1xf32>) -> tensor<5x1xf32>
```

we'll be able to get `8x2` from getResultShape, but what need here is to extract an `5x1` tensor. The interchange is not needed in this case because inner_dims_pos is [0, 1]. We'll need it in other permutation cases. This is the reason why I compute it myself.

We probably can consider about adding some utils for inner tiles though.